### PR TITLE
feat(dashboard): [#293] add training metric logging

### LINKED ISSUE #293: [Dashboard]: Add metric logging API and XML parsing for training session curves
### Summary

Introduce structured metric logging to the core logging module and extend the dashboard scanner, types, and API endpoints to parse and serve metric data for training runs.

### Scope

* Add a new public method on `FennLogger`, e.g. `log_metric(name: str, value: float, step: int) -> None`, where `name` is one of `"loss"`, `"metric"`, `"learning_rate"` - TBD
* Implement `_write_metric` XML writer in `XmlMixin` to append `<metric>` nodes to `.fn` files
* Define `MetricPoint` TypedDict and update `SessionData` and `SessionPagePayload` in `fenn/dashboard/types.py`
* Extract `<metric>` nodes during `.fn` file parsing in `fenn/dashboard/scanner.py`
* Ensure `metrics` flow through `/api/session/<project>/<session_id>` while stripping them from the `/api/sessions` listing in `fenn/dashboard/app.py`

### Acceptance Criteria

* Calling `logger.log_metric()` appends valid, safely escaped `<metric>` XML elements to `.fn` files
* `FennScanner` correctly parses `<metric>` nodes into `SessionData.metrics`
* Backward compatibility is preserved for existing `.fn` files without metrics (`metrics == []`)
* The session detail API includes metric data while the lightweight sessions list API strips it

### Parent Issue

#290

### PR DESCRIPTION
**Pull request by AaronProbha18**
Fixes: #293

## Changes
* Added `FennLogger.log_metric(name: str, value: float, step: int) -> None` to `fenn/logging.py` for structured metric emission during training runs. `name` accepts any string and is persisted as-is (no fixed enum) — the docstring was corrected to reflect this after confirming the actual names fenn's built-in trainers produce (`train_loss`, `val_loss`, `acc`), rather than the originally-assumed `loss`/`metric`/`learning_rate`.
* `log_metric()` no-ops (returns silently) if the logger has not been configured yet — i.e. `write_config()` was never called, so `fn_file` doesn't exist as an attribute. Mirrors the existing tolerance in `FennHandler.emit()` (`if self._fn_xml and self._log_file:`) for unconfigured loggers.
* Added `XmlMixin._write_metric()` to `fenn/logging.py`, appending self-closing `<metric name="..." step="..." value="..." ts="..." />` elements to `.fn` files, using the same atomic-append + escaping pattern as `_write_entry()`.
* Added `MetricPoint` `TypedDict` to `fenn/dashboard/types.py` (`name: str`, `step: int`, `value: float`, `ts: str`).
* Added `metrics: List[MetricPoint]` to `SessionData` and `SessionPagePayload` in `fenn/dashboard/types.py`. Intentionally **not** added to `SessionListItem`.
* Extended `FennScanner._parse_uncached()` in `fenn/dashboard/scanner.py` to parse `<metric>` nodes into `SessionData.metrics`, skipping elements with a missing `name` or non-numeric `step`/`value` rather than raising.
* Wired `log_metric()` calls into `ClassificationTrainer.fit()`, `RegressionTrainer.fit()`, and `LoRATrainer.fit()` (`fenn/nn/trainers/`) so all three trainers emit `train_loss` every epoch, and `val_loss`/`acc` on validation epochs — no user code changes required to get metrics into the dashboard when using these trainers.

## Notes
* No changes were needed in `fenn/dashboard/app.py`. `api_session()` already returns the full parsed session dict via `jsonify(data)`; `api_sessions()` already delegates entirely to `scanner.list_sessions()`, where the strip is implemented.
* Backward compatibility confirmed: `.fn` files with no `<metric>` elements parse with `metrics == []`.

### CODE REVIEW

File: fenn/logging.py [hunk 1]
```diff
@@ -172,6 +183,17 @@ def write_config(self, args: dict[str, Any], config_file: Path) -> None:
...
 
+    def log_metric(self, name: str, value: float, step: int) -> None:
+        """Appends a structured `<metric>` entry to the run's `.fn` file.
+
+        `name` accepts any string and is persisted as-is (e.g. "train_loss",
+        "val_loss", "acc" - the values fenn's built-in trainers emit).
+        """
+        fn_file = getattr(self, "fn_file", None)
+        if fn_file is None:
+            return
```
Comment (ApusBerliozi) [inline fenn/logging.py]: I think we can remove temp `fn_file` here, and write it like  `if getattr(self, "fn_file", None) is None`

File: fenn/dashboard/scanner.py [hunk 2]
```diff
@@ -361,6 +362,25 @@ def _parse_uncached(path: Path, stat: os.stat_result) -> SessionData | None:
...
+                continue
+            metrics.append(
+                MetricPoint(
+                    name=name,
+                    step=step,
+                    value=value,
+                    ts=metric.get("ts", ""),
+                )
+            )
+
```
Comment (ApusBerliozi) [inline fenn/dashboard/scanner.py]: I would suggest to extract this logic into separate method. Since `_parse_uncached()` is already big, it's harder to navigate inside it. So please incapsulate metrics logic inside `_get_metrics()`, so we have something like this:
`metrics = self._get_metrics(root)`

Also, it isn't really necessary to add annotation to each variable. Usually it is enough to make contracts typed

File: fenn/logging.py (hunk 1 above)
Comment (AaronProbha18) [reply fenn/logging.py]: Yep! I just removed the temporary variable. The `getattr(self, "fn_file", None) is None` check is now done directly in the `if`, and `self.fn_file` is used directly in the `_write_metric` call.

File: fenn/dashboard/scanner.py (hunk 2 above)
Comment (AaronProbha18) [reply fenn/dashboard/scanner.py]: Done! `_parse_uncached()` now just delegates to `FennScanner._get_metrics(root)`, so all the parsing/skip logic lives in a dedicated `_get_metrics(root) -> list[MetricPoint]` static method right below it.

**Review by ApusBerliozi — APPROVED** [review]

**Comment by ApusBerliozi** [comment]
Thx 4 quick response @AaronProbha18 :)

Merged
